### ISSUE TITLE
Non deterministic test output from a pre trained model (that has no Dropout Layers)

### ISSUE THREAD

**Issue by aravindsrinivas** [BVLC/caffe#3055]
Hi,

I have been trying to deploy a model that I obtained on training a deep network. The deep network has LMDB source, in a different format (each sample ties 4 gray scale images together). While deploying the network, I just need the output for a single entity (ie , a single cv::Mat with 4 gray scale images). I converted that single sample to an LMDB file, and ran a deploy.prototxt with caffe test tool from CAFFE_ROOT/build. However, I find that the argmax output, is non deterministic. That is, across multiple runs, I find different values of argmax. 

I searched for similar issues, and found that people had Dropout layers while training and were advised to remove it for test phase, and also remove all the learning and weight decay parameters of layers in the deploy.prototxt file. I have done the same too. Infact, my train prototxt doesn't have Dropout layers.

I couldn't see the blob values in the test tool, so I ran the same LMDB source on caffe train tool on bash, and just observed the output of a single forward pass. What I noticed is, the single LMDB sample I have, gives different values for the top_blob from the data layer I use, in different runs, as a result of which the top blob mean values from each layer, differ across multiple runs of the same LMDB source (I am seeing only the first forward pass's output across multiple runs). This, in a way, explains why the outputs are also different, because the blob data of each layer, right from the top blob from the input data layer, is different. I do not understand why this happens: When I have only a single image in the LMDB, and specify batch 1, it has to be a deterministic input to the network right?

Can someone point me to why this non determinism happens? There is a big need for a pre trained model to work deterministically for my task.

I am copying my deploy.prototxt and the prototxt file I used for training below:

``` bash
name: "AtariDeepNetwork"
layer {
  name: "data"
  type: "Data"
  top: "data"
  include {
    phase: TEST
  }
  data_param {
    source: "./build/examples/cpp_classification/atari_lmdb_seaquest_deploy/atari_deploy_lmdb"
    backend: LMDB
    scale : 0.0078125
    batch_size: 1
  }
  transform_param {
    mean_file : "/scratch/RLProject/Atari_Caffe/caffe-atari/atari_lmdb_seaquest/atari_mean.binaryproto"
  }
}
layer {
  name: "conv_1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  convolution_param {
    num_output: 16
    kernel_size: 8
    stride: 4
  }
}

layer {
  name: "tanh_1"
  type: "TanH"
  bottom: "conv1"
  top: "tanh1"
}


layer {
  name: "conv2"
  type: "Convolution"
  bottom: "tanh1"
  top: "conv2"
  convolution_param {
    num_output: 32
    kernel_size: 4
    stride: 2
  }
}

layer {
  name: "tanh2"
  type: "TanH"
  bottom: "conv2"
  top: "tanh2"
}

layer {
  name: "fully_connected_1"
  type: "InnerProduct"
  inner_product_param {
    num_output: 256
  }
  bottom: "tanh2"
  top: "fully_connected1"
}


layer {
  name: "relu"
  type: "ReLU"
  bottom: "fully_connected1"
  top: "relu_output"
}

layer {
  name: "fully_connected_2"
  type: "InnerProduct"
  inner_product_param {
  num_output: 18
  }
  bottom: "relu_output"
  top: "fully_connected2"
}

layer {
  name: "prob"
  type: "Softmax"
  bottom: "fully_connected2"
  top: "prob"
}
layer {
  name: "argmax"
  type: "ArgMax"
  bottom: "prob"
  top: "argmax"
  argmax_param {
    top_k: 1
   }
  include {
    phase: TEST
  }
}
```

Train_prototxt:

``` bash
name: "AtariDeepNetwork"
layer {
  name: "data"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  data_param {
    source: "/scratch/RLProject/Atari_Caffe/caffe-atari/atari_lmdb_seaquest/atari_train_lmdb"
    backend: LMDB
    scale : 0.0078125
    batch_size: 32
  }
  transform_param {
    mean_file : "/scratch/RLProject/Atari_Caffe/caffe-atari/atari_lmdb_seaquest/atari_mean.binaryproto"
  }
}

layer {
  name: "conv_1"
  type: "Convolution"
  param { lr_mult: 1}
  param { lr_mult: 2}
  bottom: "data"
  top: "conv1"
  convolution_param {
    num_output: 16
    kernel_size: 8
    stride: 4
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}

layer {
  name: "tanh_1"
  type: "TanH"
  bottom: "conv1"
  top: "tanh1"
}
layer {
  name: "conv2"
  type: "Convolution"
  param {lr_mult: 1}
  param {lr_mult: 2}
  bottom: "tanh1"
  top: "conv2"
  convolution_param {
    num_output: 32
    kernel_size: 4
    stride: 2
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "tanh2"
  type: "TanH"
  bottom: "conv2"
  top: "tanh2"
}

layer {
  name: "fully_connected_1"
  type: "InnerProduct"
  param { lr_mult: 1 decay_mult: 1 }
  param { lr_mult: 2 decay_mult: 0 }
  inner_product_param {
    num_output: 256
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
  bottom: "tanh2"
  top: "fully_connected1"
}
layer {
  name: "relu"
  type: "ReLU"
  bottom: "fully_connected1"
  top: "relu_output"
}

layer {
  name: "fully_connected_2"
  type: "InnerProduct"
  param { lr_mult: 1 decay_mult: 1 }
  param { lr_mult: 2 decay_mult: 0 }
  inner_product_param {
  num_output: 18
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
  bottom: "relu_output"
  top: "fully_connected2"
}

layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "fully_connected2"
  bottom: "label"
  top: "loss"
  include {
    phase: TRAIN
  }
}
layer {
 name: "accuracy"
 type: "Accuracy"
 bottom: "fully_connected2"
 bottom: "label"
 top: "accuracy"
 include {
   phase: TRAIN
 }
}

layer {
 name: "softmax"
 type: "Softmax"
 bottom: "fully_connected2"
 top: "prob"
 include {
   phase: TRAIN
 }
}
```

Also, here are some examples of what I said:

From the deploy, on running 100 times, I get these labels in the following distribution:

[(16, 61), (0, 23), (11, 12), (5, 3), (13, 1)]

That is, in 100 runs, the label 16 was given 61 % of the times, label 0, 23% , so on.

In a forward pass of the train, I get argmax outputs of 16, 0, 16 , .. (similar to the above), and the top blob from the data layer had values of 3.7, 3.63, 3.5, 3.65, .. (different in each run)...

Would be great if someone can give ideas to understand what is going wrong.

**Comment by seanbell** [comment]
Does the LMDB contain just one image?  The LMDB layer should cycle through all the images each forward pass.

What OS and GPU are you using, and are you using cuDNN?

**Comment by aravindsrinivas** [comment]
@seanbell 

Yes, the script for LMDB conversion prints a log message saying "Processed 1 file". 

Extract out the data blob and make sure it is always the same - Do you mean I would have to force the data blob to take a particular value?

I am using Ubuntu 12.04. I have been running on CPU mode.

**Comment by seanbell** [comment]
If you're on CPU mode, then all other layers should be deterministic (unless you found a bug).  The main thing you haven't verified is that the data blob is not changing each iteration.  

Are you on the newest master branch?  The DataLayer includes data augmentation, and in the past there was a time when some data augmentation was always enabled during TEST mode (#1430).

**Comment by ronghanghu** [comment]
An note on this: on CPU mode can be non-deterministic if you build Caffe with Intel MKL. Details in https://github.com/BVLC/caffe/issues/3109#issuecomment-146280655

To get deterministic results with MKL, set `export MKL_CBWR=AUTO`.

**Comment by temak** [comment]
@aravindsrinivas , I also observe undeterministic output from the netwrok when try to deploy it and run on test dataset, using batch_size=2.  But I use caffe in GPU mode. Have you solved the issue? Any ideas?

**Comment by shelhamer** [comment]
Nondeterminism in CPU mode is possible with MKL, and possible in GPU mode with a faulty driver version/wrong arch compilation/and in rare cases due to cuDNN.
